### ISSUE TITLE
"Bad" images cause segfault

### ISSUE THREAD

**Issue by cohoe** [jocosocial/swiftarr#52]
We noticed that sometimes people uploaded a weird image that libgd did not understand. The runtime then segfaulted. IIRC at least one of these was a TIFF file.

**Comment by cohoe** [comment]
https://stackoverflow.com/questions/46211985/different-image-file-format-samples

**Comment by cohoe** [comment]
Potential link to our image woes: https://github.com/jocosocial/swiftarr/commit/a1877b76863f6d477efc98f12bbe8c7454707126

**Comment by cohoe** [comment]
Logs:

```
web_1         | [ WARNING ] stream ended at an unexpected time [request-id: BF80741F-48B2-49B9-B247-5035B4B90162]
web_1         | GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x52 0x49GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50/run.sh: line 28:     9 Segmentation fault      (core dumped) /app/Run serve --env "${ENVIRONMENT}" --hostname ${hostname} --port ${PORT}
web_1         | Swiftarr start.
```

```
web_1         | [ WARNING ] ErrorResponse(error: true, status: 400, reason: "post text cannot be empty.", fieldErrors: Optional(["text": "post text cannot be empty."])) [request-id: FED2D76D-EF91-4BE6-83F0-C5AED544FA90]
web_1         | GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x47 0x49GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50/run.sh: line 28:     9 Segmentation fault      (core dumped) /app/Run serve --env "${ENVIRONMENT}" --hostname ${hostname} --port ${PORT}
web_1         | Swiftarr start.
```

**Comment by cohoe** [comment]
Per https://en.wikipedia.org/wiki/List_of_file_signatures the 0x89 0x50 magic is a PNG. I attempted to upload a PNG renamed as a JPG but it didn't crash. Didn't even generate an error. Uploading a BMP did but not a segfault. 

Maybe related: https://github.com/jocosocial/swiftarr/blob/506f07ee4554116fdc6edb14b085d67fea4dcd02/Sources/gdOverrides/gd_jpeg_custom.c#L262

**Comment by cohoe** [comment]
Uploading a BMP:
```
GD Warning: Cannot open TIFF imageGD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x42 0x4dGD Warning: gd-webp cannot get webp info: Not a TIFF or MDI file, bad magic number 19778 (0x4d42).
[ WARNING ] invalidImage(reason: "No matching raster formatter for given image found") [request-id: 9CC987EE-242D-462C-A012-D828EBC7759F]
```

Uploading a TIFF:
Fails to render in the browser, but the file is OK.

Uploading a HEIC:
```
GD Warning: Cannot open TIFF imageGD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x49 0x49GD Warning: gd-webp cannot get webp infoGD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x00 0x00GD Warning: gd-webp cannot get webp info: Not a TIFF or MDI file, bad magic number 0 (0x0).

GD Warning: Cannot open TIFF imageGD Warning: one parameter to a memory allocation multiplication is negative or zero, failing operation gracefully

[ WARNING ] invalidImage(reason: "No matching raster formatter for given image found") [request-id: 39296EFC-2356-49F7-8CCB-A4A05A6F3572]
```

Except when I do it again.
```
GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x00 0x00GD Warning: gd-webp cannot get webp info: Not a TIFF or MDI file, bad magic number 0 (0x0).

GD Warning: Cannot open TIFF imageGD Warning: one parameter to a memory allocation multiplication is negative or zero, failing operation gracefully

[ WARNING ] invalidImage(reason: "No matching raster formatter for given image found") [request-id: F2B1667E-1D16-47E5-97B0-52DD8080F7D9]
```

I wonder.....

AH HA! If I upload an OK file (PNG or JPG) beforehand, then make a request with a HEIC, it blows up but keeps the "starts with 0x..." of the previous. If you upload again it goes to "starts with 0x00 0x00". This smells of memory corruption/access. The warning that it's generating about allocation here may be relevant in that this is currently not segfaulting but in a multi-user environment it might be.

**Comment by cohoe** [comment]
tldr: Upload a HEIC and GD explodes, likely memory corruption. Do this in a multi-threaded way and it's entirely possible that produces the other segfault we observed. Easiest thing might be to just block HEIC uploads. But idk what the default iOS format is these days and if that would make people sad.

**Comment by cohoe** [comment]
Closing until it comes up again.

**Comment by cohoe** [comment]
Came up at least once in 2025. Last log message was:
```
GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50GD Warning: gd-jpeg: JPEG library reports unrecoverable error: Not a JPEG file: starts with 0x89 0x50
```
Only seemed to happen embarkation day. Forgot to grab the log backup before shutting down.

**Comment by cohoe** [comment]
I'm really hoping that the switch to vips from gd fixes this
